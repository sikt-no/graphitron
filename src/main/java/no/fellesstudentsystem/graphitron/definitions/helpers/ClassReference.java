package no.fellesstudentsystem.graphitron.definitions.helpers;

import com.squareup.javapoet.ClassName;
import no.fellesstudentsystem.graphitron.configuration.GeneratorConfig;
import no.fellesstudentsystem.graphitron.configuration.externalreferences.CodeReference;

public class ClassReference {
    private final Class<?> classReference;
    private final ClassName className;

    public ClassReference(CodeReference reference) {
        if (reference != null) {
            var recordClass = GeneratorConfig.getExternalReferences().getClassFrom(reference);
            classReference = recordClass;
            className = ClassName.get(recordClass);
        } else {
            classReference = null;
            className = null;
        }
    }

    public ClassReference(String typeName, String packageSource) {
        Class<?> reference;
        try {
            reference = Class.forName(packageSource + "." + typeName);
        } catch (ClassNotFoundException e) {
            reference = null;
        }
        classReference = reference;
        className = classReference != null
                ? ClassName.get(classReference)
                : ClassName.get(packageSource, typeName); // TODO: Remove special else case (should be just null). In our test the objects do not exist and this class lookup will fail. When the graphql-codegen is integrated in this module this will be redundant.
    }

    public Class<?> getReferenceClass() {
        return classReference;
    }

    public String getClassNameString() {
        return className != null ? className.simpleName() : "";
    }

    public ClassName getClassName() {
        return className;
    }
}
