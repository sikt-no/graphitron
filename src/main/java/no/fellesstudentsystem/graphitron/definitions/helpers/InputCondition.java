package no.fellesstudentsystem.graphitron.definitions.helpers;

import com.squareup.javapoet.CodeBlock;
import no.fellesstudentsystem.graphitron.definitions.fields.InputField;

import java.util.LinkedHashSet;
import java.util.stream.Collectors;

public class InputCondition {
    private final InputField input;
    private final String namePath;
    private final LinkedHashSet<String> nullChecks;
    private final boolean pastWasIterable;

    private InputCondition(InputField input, String namePath, LinkedHashSet<String> nullChecks, boolean pastWasIterable) {
        this.input = input;
        this.namePath = namePath;
        this.nullChecks = new LinkedHashSet<>(nullChecks);
        this.pastWasIterable = pastWasIterable;

        inferAdditionalChecks(input);
    }

    public InputCondition(InputField input) {
        this(input, "",  new LinkedHashSet<>(), false);
    }

    public InputField getInput() {
        return input;
    }

    private void inferAdditionalChecks(InputField input) {
        var name = getNameWithPathString();
        if (!pastWasIterable && input.isNullable()) {
            this.nullChecks.add(name + " != null");
        }

        if (input.isIterableWrapped()) {
            this.nullChecks.add(name + ".size() > 0");
        }
    }

    public String getNamePath() {
        return namePath;
    }

    public CodeBlock getNameWithPath() {
        return CodeBlock.of(getNameWithPathString());
    }

    public String getNameWithPathString() {
        return namePath.isEmpty() ? input.getName() : namePath + input.getMappingFromSchemaName().asGetCall().toString();
    }

    public String getChecksAsSequence() {
        return !nullChecks.isEmpty() ? nullChecks.stream().sorted().collect(Collectors.joining(" && ")) : "";
    }

    public InputCondition iterate(InputField input) {
        return new InputCondition(input, getNameWithPathString(), nullChecks, pastWasIterable || this.input.isIterableWrapped());
    }

    public InputCondition applyTo(InputField input) {
        return new InputCondition(input, namePath, nullChecks, pastWasIterable || this.input.isIterableWrapped());
    }
}
