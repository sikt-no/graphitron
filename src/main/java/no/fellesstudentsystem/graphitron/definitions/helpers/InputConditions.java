package no.fellesstudentsystem.graphitron.definitions.helpers;

import java.util.List;

public class InputConditions {
    private final List<InputCondition> independentConditions;
    private final List<ConditionTuple> conditionTuples;

    public InputConditions(List<InputCondition> independentConditions, List<ConditionTuple> conditionTuples) {
        this.independentConditions = independentConditions;
        this.conditionTuples = conditionTuples;
    }

    public List<InputCondition> getIndependentConditions() {
        return independentConditions;
    }

    public List<ConditionTuple> getConditionTuples() {
        return conditionTuples;
    }

    public static class ConditionTuple {
        private final String path;
        private final List<InputCondition> conditions;

        public ConditionTuple(String path, List<InputCondition> conditions) {
            this.path = path;
            this.conditions = conditions;
        }

        public String getPath() {
            return path;
        }

        public List<InputCondition> getConditions() {
            return conditions;
        }
    }
}
