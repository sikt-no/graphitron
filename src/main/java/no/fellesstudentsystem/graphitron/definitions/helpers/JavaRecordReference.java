package no.fellesstudentsystem.graphitron.definitions.helpers;

import com.squareup.javapoet.ClassName;
import com.squareup.javapoet.TypeName;
import no.fellesstudentsystem.graphitron.configuration.GeneratorConfig;
import no.fellesstudentsystem.graphitron.configuration.externalreferences.CodeReference;

public class JavaRecordReference {
    private final Class<?> classReference;
    private final TypeName typeName;

    public JavaRecordReference(CodeReference reference ) {
        if (reference != null) {
            var recordClass = GeneratorConfig.getExternalReferences().getClassFrom(reference);
            this.classReference = recordClass;
            this.typeName = ClassName.get(recordClass);
        } else {
            this.classReference = null;
            this.typeName = null;
        }
    }

    public Class<?> getRecordClass() {
        return classReference;
    }

    public String getClassName() {
        return classReference != null ? classReference.getSimpleName() : "";
    }

    public TypeName getTypeName() {
        return typeName;
    }
}
