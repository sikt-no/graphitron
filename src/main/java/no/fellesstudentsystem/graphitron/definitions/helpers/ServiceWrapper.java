package no.fellesstudentsystem.graphitron.definitions.helpers;

import com.squareup.javapoet.ClassName;
import com.squareup.javapoet.TypeName;
import no.fellesstudentsystem.graphitron.configuration.GeneratorConfig;
import no.fellesstudentsystem.graphitron.definitions.fields.ObjectField;
import no.fellesstudentsystem.graphql.schema.ProcessedSchema;

import java.lang.reflect.Method;
import java.lang.reflect.ParameterizedType;
import java.lang.reflect.Type;

import static no.fellesstudentsystem.graphitron.generators.codebuilding.ClassNameFormat.wrapListIf;

/**
 * Class that contains extended information about a mutation service.
 */
public class ServiceWrapper {
    private final boolean returnsJavaRecord;
    private final Class<?> service, returnType;
    private final Method method;
    private final TypeName returnTypeName;
    private final ClassName serviceClassName;

    public ServiceWrapper(ObjectField field, ProcessedSchema processedSchema) {
        var reference = field.getServiceReference();
        var references = GeneratorConfig.getExternalReferences();
        service = references.getClassFrom(reference);
        method = references.getNullableMethodFrom(reference);
        serviceClassName = ClassName.get(service);

        var response = processedSchema.getObject(field);
        returnsJavaRecord = response != null && response.hasJavaRecordReference();
        if (returnsJavaRecord) {
            returnType = response.getRecordReference();
            returnTypeName = response.getRecordClassName();
        } else {
            returnType = method != null ? extractType(method.getGenericReturnType()) : null;
            returnTypeName = (returnType != null) ? wrapListIf(ClassName.get(returnType), field.isIterableWrapped()) : null;
        }
    }

    /**
     * @return If the type is a parameterized type such as a List or Set, return the class of the parameter, otherwise return the class of the argument.
     */
    public static Class<?> extractType(Type type) {
        if (type == null) return null;
        return (Class<?>) (type instanceof ParameterizedType ? ((ParameterizedType) type).getActualTypeArguments()[0] : type);
    }

    /**
     * @return Is the return type of this method located within this method's service class? Used for special return classes.
     */
    public boolean returnsJavaRecord() {
        return returnsJavaRecord;
    }

    /**
     * @return The class of the return type for this service method.
     */
    public Class<?> getReturnType() {
        return returnType;
    }

    /**
     * @return The reflected method object for this service method.
     */
    public Method getMethod() {
        return method;
    }

    /**
     * @return Reference of the class that the service method is located in.
     */
    public Class<?> getService() {
        return service;
    }

    /**
     * @return Javapoet classname of the service.
     */
    public ClassName getServiceClassName() {
        return serviceClassName;
    }

    /**
     * @return Javapoet classname of the return type.
     */
    public TypeName getReturnTypeName() {
        return returnTypeName;
    }
}
