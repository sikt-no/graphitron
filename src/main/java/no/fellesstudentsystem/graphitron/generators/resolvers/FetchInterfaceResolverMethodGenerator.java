package no.fellesstudentsystem.graphitron.generators.resolvers;

import com.squareup.javapoet.CodeBlock;
import com.squareup.javapoet.MethodSpec;
import com.squareup.javapoet.TypeName;
import no.fellesstudentsystem.graphitron.definitions.fields.ObjectField;
import no.fellesstudentsystem.graphitron.definitions.objects.AbstractObjectDefinition;
import no.fellesstudentsystem.graphitron.definitions.objects.InterfaceDefinition;
import no.fellesstudentsystem.graphitron.definitions.objects.ObjectDefinition;
import no.fellesstudentsystem.graphitron.generators.abstractions.ResolverMethodGenerator;
import no.fellesstudentsystem.graphitron.generators.dependencies.Dependency;
import no.fellesstudentsystem.graphitron.generators.dependencies.QueryDependency;
import no.fellesstudentsystem.graphitron.schema.ProcessedSchema;
import no.fellesstudentsystem.graphitron.generators.abstractions.DBClassGenerator;

import java.util.Comparator;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

import static no.fellesstudentsystem.graphitron.generators.db.FetchDBClassGenerator.SAVE_DIRECTORY_NAME;
import static no.fellesstudentsystem.graphitron.mappings.JavaPoetClassName.*;
import static org.apache.commons.lang3.StringUtils.capitalize;
import static org.apache.commons.lang3.StringUtils.uncapitalize;

/**
 * Generates resolvers for queries returning an interface. E.g. the node resolver.
 */
public class FetchInterfaceResolverMethodGenerator extends ResolverMethodGenerator<ObjectField> {
    private final ObjectDefinition localObject;

    public FetchInterfaceResolverMethodGenerator(ObjectDefinition localObject, ProcessedSchema processedSchema) {
        super(localObject, processedSchema);
        this.localObject = localObject;
    }

    @Override
    public MethodSpec generate(ObjectField target) {
        InterfaceDefinition interfaceDefinition = processedSchema.getInterface(target);
        TypeName returnClassName = interfaceDefinition.getGraphClassName();

        var spec = getDefaultSpecBuilder(target.getName(), returnClassName);

        if (!localObject.isRoot()) {
            spec.addParameter(localObject.getGraphClassName(), uncapitalize(localObject.getName()));
        }

        var inputField = target.getInputFields().get(0);
        String inputFieldName = inputField.getName();
        spec
                .addParameter(inputIterableWrap(inputField), inputFieldName)
                .addParameter(DATA_FETCHING_ENVIRONMENT.className, ENV_NAME)
                .addStatement("String tablePartOfId = $T.getTablePartOf($N)", FIELD_HELPERS.className, inputFieldName)
                .addCode("\n");

        processedSchema
                .getObjects()
                .values()
                .stream()
                .filter(it -> it.implementsInterface(interfaceDefinition.getName()))
                .sorted(Comparator.comparing(AbstractObjectDefinition::getName))
                .forEach(implementation -> {
                    var queryLocation = implementation.getName() + DBClassGenerator.FILE_NAME_SUFFIX;
                    var map = Map.ofEntries(
                            Map.entry("env", ENV_NAME),
                            Map.entry("string", STRING.className),
                            Map.entry("returnType", returnClassName),
                            Map.entry("dataloader", DATA_LOADER_FACTORY.className),
                            Map.entry("batchLoader", MAPPED_BATCH_LOADER_WITH_CONTEXT.className),
                            Map.entry("implementationClass", implementation.getGraphClassName()),
                            Map.entry("future", COMPLETABLE_FUTURE.className),
                            Map.entry("queryInstanceField", uncapitalize(queryLocation)),
                            Map.entry("implName", implementation.getName()),
                            Map.entry("inputFieldName", inputFieldName),
                            Map.entry("inputFieldNameCap", capitalize(inputFieldName)),
                            Map.entry("selectionSets", SELECTION_SET.className),
                            Map.entry("environmentUtils", ENVIRONMENT_UTILS.className),
                            Map.entry("referenceFieldName", capitalize(target.getName())),
                            Map.entry("context", Dependency.CONTEXT_NAME)
                    );
                    dependencySet.add(new QueryDependency(queryLocation, SAVE_DIRECTORY_NAME));

                    spec.beginControlFlow("if (tablePartOfId.equals($T.$N.getViewId().toString()))", TABLES.className, implementation.getTable().getMappingName());
                    spec.addStatement(CodeBlock.builder().addNamed(
                                    "return $env:N.getDataLoaderRegistry().<$string:T, $returnType:T>computeIfAbsent(tablePartOfId, name ->$W" +
                                            "$dataloader:T.newMappedDataLoader(($batchLoader:T<$string:T, $implementationClass:T>) (keys, loaderEnvironment) ->$>$W" +
                                            "$future:T.completedFuture($queryInstanceField:N.load$implName:NBy$inputFieldNameCap:NsAs$referenceFieldName:N" +
                                            "($context:N, keys, new $selectionSets:T($environmentUtils:T.getSelectionSetsFromEnvironment(loaderEnvironment))))))" +
                                            "$<$W.load($inputFieldName:N, $env:N)", map)
                            .build());
                    spec.endControlFlow();
                });

        spec.addStatement("throw new $T(\"could not find dataloader for $N with prefix \" + tablePartOfId)", ILLEGAL_ARGUMENT_EXCEPTION.className, inputFieldName);

        return spec.build();
    }

    @Override
    public List<MethodSpec> generateAll() {
        return localObject
                .getFields()
                .stream()
                .filter(ObjectField::isGenerated)
                .filter(it -> processedSchema.isInterface(it.getTypeName()))
                .map(this::generate)
                .collect(Collectors.toList());
    }

    @Override
    public boolean generatesAll() {
        return localObject
                .getFields()
                .stream()
                .filter(it -> processedSchema.isInterface(it.getTypeName()))
                .allMatch(ObjectField::isGenerated);
    }
}
