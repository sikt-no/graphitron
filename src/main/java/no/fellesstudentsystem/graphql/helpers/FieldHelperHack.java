package no.fellesstudentsystem.graphql.helpers;

import java.nio.charset.StandardCharsets;
import java.util.*;

/**
 * A helper class that handles some special ID-unpacking. This does not belong in Graphitron and should be removed at some point.
 */
@Deprecated
public class FieldHelperHack {
    /**
     * @param packageName Encoded ID package name
     * @return The table contained within the ID package name
     * @throws IllegalArgumentException If the encoded ID package is not a valid ID.
     */
    public static String getTablePartOf(String packageName) {
        return getTablePartOf(packageName, decodeIDPackage(packageName));
    }

    private static String getTablePartOf(String packageName, String id) {
        if (id.indexOf(':') < 0) {
            throw new IllegalArgumentException(String.format("%s (%s) is not a valid ID.", packageName, id));
        }
        return id.substring(0, id.indexOf(':'));
    }

    private static String decodeIDPackage(String s) {
        try {
            return new String(Base64.getUrlDecoder().decode(s), StandardCharsets.UTF_8);
        } catch (IllegalArgumentException e) {
            throw new IllegalArgumentException(s + " is not a valid ID.");
        }
    }
}
