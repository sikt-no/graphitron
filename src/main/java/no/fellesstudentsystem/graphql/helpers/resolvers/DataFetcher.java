package no.fellesstudentsystem.graphql.helpers.resolvers;

import graphql.ErrorType;
import graphql.GraphqlErrorBuilder;
import graphql.schema.DataFetchingEnvironment;
import no.fellesstudentsystem.graphql.exception.ValidationViolationGraphQLException;
import no.fellesstudentsystem.graphql.helpers.EnvironmentUtils;
import no.fellesstudentsystem.graphql.helpers.functions.DBCount;
import no.fellesstudentsystem.graphql.helpers.functions.DBQuery;
import no.fellesstudentsystem.graphql.helpers.functions.DBQueryRoot;
import no.fellesstudentsystem.graphql.helpers.selection.SelectionSet;
import no.fellesstudentsystem.graphql.relay.ConnectionImpl;
import org.apache.commons.lang3.tuple.Pair;
import org.dataloader.DataLoaderFactory;
import org.dataloader.MappedBatchLoaderWithContext;
import org.jooq.DSLContext;

import java.util.*;
import java.util.concurrent.CompletableFuture;
import java.util.function.Function;
import java.util.stream.Collectors;

import static no.fellesstudentsystem.graphql.naming.GraphQLReservedName.CONNECTION_TOTAL_COUNT;

public class DataFetcher extends AbstractFetcher {
    /**
     * Create a dataloader for a resolver.
     * @param env Environment for the resolver.
     * @param ctx Context from the resolver.
     */
    public DataFetcher(DataFetchingEnvironment env, DSLContext ctx) {
        super(env, ctx);
    }

    /**
     * Load the data for a root resolver.
     * @param dbFunction Function to call to retrieve the query data.
     * @return A resolver result.
     * @param <T> Type that the resolver fetches.
     */
    public <T> CompletableFuture<T> load(DBQueryRoot<T> dbFunction) {
        return CompletableFuture.completedFuture(dbFunction.callDBMethod(this.ctx, selection));
    }

    /**
     * Load the data for a root resolver. The result is paginated.
     * @param pageSize Size of the pages for pagination.
     * @param maxNodes Limit on how many elements may be fetched at once.
     * @param dbFunction Function to call to retrieve the query data.
     * @param countFunction Function to call to retrieve the total count of elements that could be potentially retrieved.
     * @param connectionFunction Function that converts the result of the query to a GraphQL connection structure.
     * @return A paginated resolver result.
     * @param <T> Type that the resolver fetches.
     */
    public <T, U> CompletableFuture<U> loadPaginated(
            int pageSize,
            int maxNodes,
            DBQueryRoot<List<Pair<String, T>>> dbFunction,
            DBCount<String> countFunction,
            Function<ConnectionImpl<T>, U> connectionFunction
    ) {
        return CompletableFuture.completedFuture(
                getPaginatedConnection(
                        dbFunction.callDBMethod(ctx, connectionSelection),
                        pageSize,
                        connectionSelection.contains(CONNECTION_TOTAL_COUNT.getName()) ? countFunction.callDBMethod(ctx, Set.of()) : -1,
                        maxNodes,
                        connectionFunction
                )
        );
    }

    /**
     * Load the data for a resolver.
     * @param resolveName Name of the resolver.
     * @param dbFunction Function to call to retrieve the query data.
     * @param id ID of the queried element.
     * @return A resolver result.
     * @param <T> Type that the resolver fetches.
     */
    public <T> CompletableFuture<T> load(String resolveName, String id, DBQuery<String, T> dbFunction) {
        return getLoader(resolveName, (keys, set) -> getMappedDataLoader(keys, set, dbFunction))
                .load(asKeyPath(id), env);
    }

    /**
     * Load the data for a resolver. The result is paginated.
     * @param pageSize Size of the pages for pagination.
     * @param maxNodes Limit on how many elements may be fetched at once.
     * @param dbFunction Function to call to retrieve the query data.
     * @param countFunction Function to call to retrieve the total count of elements that could be potentially retrieved.
     * @param connectionFunction Function that converts the result of the query to a GraphQL connection structure.
     * @return A paginated resolver result.
     * @param <T> Type that the resolver fetches.
     */
    public <T, U> CompletableFuture<U> loadPaginated(
            String resolveName,
            String id,
            int pageSize,
            int maxNodes,
            DBQuery<String, List<Pair<String, T>>> dbFunction,
            DBCount<String> countFunction,
            Function<ConnectionImpl<T>, U> connectionFunction
    ) {
        return getConnectionLoader(resolveName, (keys, set) -> getMappedDataLoader(keys, set, maxNodes, pageSize, dbFunction, countFunction, connectionFunction))
                .load(asKeyPath(id), env);
    }

    /**
     * Load the data for a resolver.
     * @param resolveName Name of the resolver.
     * @param dbFunction Function to call to retrieve the query data.
     * @param id ID of the queried element.
     * @return A resolver result.
     * @param <T> Type that the resolver fetches.
     */
    public <T> CompletableFuture<List<T>> loadNonNullable(String resolveName, String id, DBQuery<String, List<T>> dbFunction) {
        return getLoader(resolveName, (keys, set) -> getMappedDataLoader(keys, set, dbFunction))
                .load(asKeyPath(id), env)
                .thenApply(data -> Optional.ofNullable(data).orElse(List.of()));
    }

    /**
     * Load the data for a resolver in a lookup format.
     * @param keys The lookup keys that are necessary for match rows in the result with the original resolver input.
     * @param dbFunction Function to call to retrieve the query data.
     * @return A resolver result without pagination.
     * @param <K> The key type that this lookup uses.
     * @param <V> Type that the resolver fetches.
     */
    public <K, V> CompletableFuture<List<V>> loadLookup(List<List<K>> keys, DBQuery<String, V> dbFunction) {
        if (keys.isEmpty()) {
            return CompletableFuture.completedFuture(List.of());
        }

        var mergedKeys = mergeKeys(keys, env);

        var dbResult = dbFunction.callDBMethod(ctx, new HashSet<>(mergedKeys), selection);
        var orderedResult = mergedKeys.stream().map(dbResult::get).collect(Collectors.toList());
        return CompletableFuture.completedFuture(orderedResult);
    }

    public static <K> List<String> mergeKeys(List<List<K>> keys, DataFetchingEnvironment env) {
        if (keys.isEmpty()) {
            return List.of();
        }

        // This assumes all keys are correlated.
        var nKeys = keys.get(0).size();
        if (keys.stream().map(List::size).anyMatch(it -> it != nKeys)) {
            throw new ValidationViolationGraphQLException(
                    List.of(
                            GraphqlErrorBuilder
                                    .newError(env)
                                    .path(env.getExecutionStepInfo().getPath())
                                    .message("Keys sets have differing lengths. For this type of query, each key field is required to be an array of equal length.")
                                    .errorType(ErrorType.ValidationError)
                                    .build()
                    )
            );
        }

        var mergedKeys = new ArrayList<>(Collections.nCopies(nKeys, ""));
        for (var keyList : keys) {
            for (int i = 0; i < nKeys; i++) {
                var currentKey = mergedKeys.get(i);
                mergedKeys.set(i, (currentKey.isEmpty() ? "" : currentKey + ",") + keyList.get(i));
            }
        }

        return mergedKeys;
    }

    /**
     * Load the data for an interface resolver.
     * @param table The source from which the data should be loaded.
     * @param keyToLoad The specific key that should be loaded.
     * @param dbFunction Function to call to retrieve the query data.
     * @return A resolver result without pagination.
     * @param <U> Type that the resolver fetches.
     */
    public <T, U extends T> CompletableFuture<T> loadInterface(String table, String keyToLoad, DBQuery<String, U> dbFunction) {
        return env
                .getDataLoaderRegistry()
                .<String, T>computeIfAbsent(table, name ->
                        DataLoaderFactory.newMappedDataLoader((MappedBatchLoaderWithContext<String, U>) (ids, loaderEnvironment) ->
                                CompletableFuture.completedFuture(dbFunction.callDBMethod(ctx, ids, new SelectionSet(EnvironmentUtils.getSelectionSetsFromEnvironment(loaderEnvironment))))
                        )
                )
                .load(keyToLoad, env);
    }

    private <T, U> CompletableFuture<Map<String, U>> getMappedDataLoader(
            Set<String> keys,
            SelectionSet selectionSet,
            int maxNodes,
            int pageSize,
            DBQuery<String, List<Pair<String, T>>> dbFunction,
            DBCount<String> countFunction,
            Function<ConnectionImpl<T>, U> connectionFunction
    ) {
        if (keys.isEmpty()) {
            return CompletableFuture.completedFuture(Map.of());
        }

        var keyToId = getKeyToId(keys);
        var idSet = new HashSet<>(keyToId.values());
        return CompletableFuture.completedFuture(
                getPaginatedConnection(
                        resultAsMap(keyToId, dbFunction.callDBMethod(ctx, idSet, selectionSet)),
                        pageSize,
                        connectionSelection.contains(CONNECTION_TOTAL_COUNT.getName()) ? countFunction.callDBMethod(ctx, idSet) : -1,
                        maxNodes,
                        connectionFunction
                )
        );
    }


    private <T> CompletableFuture<Map<String, T>> getMappedDataLoader(Set<String> keys, SelectionSet selectionSet, DBQuery<String, T> dbFunction) {
        if (keys.isEmpty()) {
            return CompletableFuture.completedFuture(Map.of());
        }

        var keyToId = getKeyToId(keys);
        return CompletableFuture.completedFuture(resultAsMap(keyToId, dbFunction.callDBMethod(ctx, new HashSet<>(keyToId.values()), selectionSet)));
    }
}
