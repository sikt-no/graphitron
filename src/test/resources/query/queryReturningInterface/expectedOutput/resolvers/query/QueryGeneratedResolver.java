package fake.code.generated.resolvers.query;

import fake.code.generated.queries.query.FilmDBQueries;
import fake.code.generated.queries.query.InventoryDBQueries;
import fake.code.generated.queries.query.RentalDBQueries;
import fake.graphql.example.api.QueryResolver;
import fake.graphql.example.model.Node;
import fake.graphql.example.model.Titled;
import graphql.schema.DataFetchingEnvironment;
import java.lang.Exception;
import java.lang.IllegalArgumentException;
import java.lang.Override;
import java.lang.String;
import java.util.concurrent.CompletableFuture;
import javax.inject.Inject;
import no.fellesstudentsystem.graphql.helpers.FieldHelperHack;
import no.fellesstudentsystem.graphql.helpers.resolvers.DataFetcher;
import no.sikt.graphitron.jooq.generated.testdata.tables.Film;
import no.sikt.graphitron.jooq.generated.testdata.tables.Inventory;
import no.sikt.graphitron.jooq.generated.testdata.tables.Rental;
import org.jooq.DSLContext;

public abstract class QueryGeneratedResolver implements QueryResolver {
    @Inject
    DSLContext ctx;

    @Override
    public CompletableFuture<Node> node(String id, DataFetchingEnvironment env) throws Exception {
        String tablePartOfId = FieldHelperHack.getTablePartOf(id);

        if (tablePartOfId.equals(Film.FILM.getViewId().toString())) {
            return new DataFetcher(env, this.ctx).loadInterface(tablePartOfId, id, (ctx, ids, selectionSet) -> FilmDBQueries.loadFilmByIdsAsNode(ctx, ids, selectionSet));
        }
        if (tablePartOfId.equals(Inventory.INVENTORY.getViewId().toString())) {
            return new DataFetcher(env, this.ctx).loadInterface(tablePartOfId, id, (ctx, ids, selectionSet) -> InventoryDBQueries.loadInventoryByIdsAsNode(ctx, ids, selectionSet));
        }
        if (tablePartOfId.equals(Rental.RENTAL.getViewId().toString())) {
            return new DataFetcher(env, this.ctx).loadInterface(tablePartOfId, id, (ctx, ids, selectionSet) -> RentalDBQueries.loadRentalByIdsAsNode(ctx, ids, selectionSet));
        }
        throw new IllegalArgumentException("Could not find dataloader for id with prefix " + tablePartOfId);
    }

    @Override
    public CompletableFuture<Titled> titled(String title, DataFetchingEnvironment env) throws
            Exception {
        String tablePartOfId = FieldHelperHack.getTablePartOf(title);

        if (tablePartOfId.equals(Film.FILM.getViewId().toString())) {
            return new DataFetcher(env, this.ctx).loadInterface(tablePartOfId, title, (ctx, ids, selectionSet) -> FilmDBQueries.loadFilmByTitlesAsTitled(ctx, ids, selectionSet));
        }
        throw new IllegalArgumentException("Could not find dataloader for title with prefix " + tablePartOfId);
    }
}
