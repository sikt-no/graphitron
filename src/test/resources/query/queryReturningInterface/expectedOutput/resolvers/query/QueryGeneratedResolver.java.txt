package fake.code.generated.resolvers.query;

import fake.code.generated.queries.query.FilmDBQueries;
import fake.code.generated.queries.query.InventoryDBQueries;
import fake.code.generated.queries.query.RentalDBQueries;
import fake.graphql.example.package.api.QueryResolver;
import fake.graphql.example.package.model.Film;
import fake.graphql.example.package.model.Inventory;
import fake.graphql.example.package.model.Node;
import fake.graphql.example.package.model.Rental;
import fake.graphql.example.package.model.Titled;
import graphql.schema.DataFetchingEnvironment;
import java.lang.Exception;
import java.lang.IllegalArgumentException;
import java.lang.Override;
import java.lang.String;
import java.util.concurrent.CompletableFuture;
import javax.inject.Inject;
import no.fellesstudentsystem.graphql.helpers.EnvironmentUtils;
import no.fellesstudentsystem.graphql.helpers.FieldHelpers;
import no.fellesstudentsystem.graphql.helpers.selection.SelectionSet;
import no.sikt.graphitron.jooq.generated.testdata.Tables;
import org.dataloader.DataLoaderFactory;
import org.dataloader.MappedBatchLoaderWithContext;
import org.jooq.DSLContext;

public abstract class QueryGeneratedResolver implements QueryResolver {
    @Inject
    DSLContext ctx;

    @Inject
    private FilmDBQueries filmDBQueries;

    @Inject
    private InventoryDBQueries inventoryDBQueries;

    @Inject
    private RentalDBQueries rentalDBQueries;

    @Override
    public CompletableFuture<Node> node(String id, DataFetchingEnvironment env) throws Exception {
        var ctx = env.getLocalContext() == null ? this.ctx : (DSLContext) env.getLocalContext();
        String tablePartOfId = FieldHelpers.getTablePartOf(id);

        if (tablePartOfId.equals(Tables.FILM.getViewId().toString())) {
            return env.getDataLoaderRegistry().<String, Node>computeIfAbsent(tablePartOfId, name ->
                    DataLoaderFactory.newMappedDataLoader((MappedBatchLoaderWithContext<String, Film>) (keys, loaderEnvironment) ->
                        CompletableFuture.completedFuture(filmDBQueries.loadFilmByIdsAsNode(ctx, keys, new SelectionSet(EnvironmentUtils.getSelectionSetsFromEnvironment(loaderEnvironment))))))
                    .load(id, env);
        }
        if (tablePartOfId.equals(Tables.INVENTORY.getViewId().toString())) {
            return env.getDataLoaderRegistry().<String, Node>computeIfAbsent(tablePartOfId, name ->
                    DataLoaderFactory.newMappedDataLoader((MappedBatchLoaderWithContext<String, Inventory>) (keys, loaderEnvironment) ->
                        CompletableFuture.completedFuture(inventoryDBQueries.loadInventoryByIdsAsNode(ctx, keys, new SelectionSet(EnvironmentUtils.getSelectionSetsFromEnvironment(loaderEnvironment))))))
                    .load(id, env);
        }
        if (tablePartOfId.equals(Tables.RENTAL.getViewId().toString())) {
            return env.getDataLoaderRegistry().<String, Node>computeIfAbsent(tablePartOfId, name ->
                    DataLoaderFactory.newMappedDataLoader((MappedBatchLoaderWithContext<String, Rental>) (keys, loaderEnvironment) ->
                        CompletableFuture.completedFuture(rentalDBQueries.loadRentalByIdsAsNode(ctx, keys, new SelectionSet(EnvironmentUtils.getSelectionSetsFromEnvironment(loaderEnvironment))))))
                    .load(id, env);
        }
        throw new IllegalArgumentException("could not find dataloader for id with prefix " + tablePartOfId);
    }

    @Override
    public CompletableFuture<Titled> titled(String title, DataFetchingEnvironment env) throws
            Exception {
        var ctx = env.getLocalContext() == null ? this.ctx : (DSLContext) env.getLocalContext();
        String tablePartOfId = FieldHelpers.getTablePartOf(title);

        if (tablePartOfId.equals(Tables.FILM.getViewId().toString())) {
            return env.getDataLoaderRegistry().<String, Titled>computeIfAbsent(tablePartOfId, name ->
                    DataLoaderFactory.newMappedDataLoader((MappedBatchLoaderWithContext<String, Film>) (keys, loaderEnvironment) ->
                        CompletableFuture.completedFuture(filmDBQueries.loadFilmByTitlesAsTitled(ctx, keys, new SelectionSet(EnvironmentUtils.getSelectionSetsFromEnvironment(loaderEnvironment))))))
                    .load(title, env);
        }
        throw new IllegalArgumentException("could not find dataloader for title with prefix " + tablePartOfId);
    }
}